### PR TITLE
[BUILD] Restore Bazel flag removed from public API

### PR DESCRIPTION
**Pull request by dbolduc**
A follow up to #2679.

This `bool_flag` was a part of the public API. If we remove it, anyone using it in downstream code have broken build commands.

So instead of removing it, I think we should mark it as deprecated.

```sh
$ bazelist build --//api:with_abseil //api:api

WARNING: opentelemetry-cpp/api/BUILD:8:10: target '//api:with_abseil' is deprecated: The value of this flag is ignored. Bazel builds always use Abseil's pre-adopted `std::` types. You should remove this flag from your build command.
INFO: Analyzed target //api:api (0 packages loaded, 545 targets configured).
INFO: Found 1 target...
Target //api:api up-to-date (nothing to build)
INFO: Elapsed time: 0.574s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
```

---

On removing the now useless flag...

I do not know what OpenTelemetry's breaking change policy is. But consider something like:
- removing the flag in the next major version bump (`v2.x`). That is [Google's policy](https://opensource.google/documentation/policies/library-breaking-change#policy) (don't laugh).
- OR giving the flag a few releases in the deprecated state, so downstream projects have a wider window to remove it.

In my own repo, I would typically open a new issue and tag the code so we remember to clean it up later.
```
TODO(#XXXX) - Remove this flag when ...
bool_flag( ... )
```

I am happy to do that here. You will have to tell me when you want to remove it. (If I have convinced y'all to take this PR).

### CODE REVIEW

**Review by marcalff — APPROVED** [review]
LGTM, thanks for the fix.

**Comment by marcalff** [comment]
@keith Please take a look and comment.

I assume keeping a unused flag is ok.

**Review by owent — APPROVED** [review]
LGTM, thanks.

**Comment by keith** [comment]
Keeping and marking deprecated is definitely on. It might be a bit misleading if folks expect it to do something but I definitely understand the issue.

File: api/BUILD [hunk 1]
```diff
@@ -55,3 +55,9 @@ config_setting(
     constraint_values = ["@bazel_tools//tools/cpp:msvc"],
     flag_values = {":with_cxx_stdlib": "best"},
 )
+
+bool_flag(
+    name = "with_abseil",
+    build_setting_default = False,
+    deprecation = "The value of this flag is ignored. Bazel builds always depend on Abseil for its pre-adopted `std::` types. You should remove this flag from your build command."
```
Comment (marcalff) [inline api/BUILD:62]:
```suggestion
    deprecation = "The value of this flag is ignored. Bazel builds always depend on Abseil for its pre-adopted `std::` types. You should remove this flag from your build command.",
```

**Pull request merged**